server/service/async/async.go

```diff
@@ -0,0 +1,280 @@
+	// rows in the VALUES section (short of doing LOAD FILE, which we can't):
+	// https://www.databasejournal.com/features/mysql/optimize-mysql-inserts-using-batch-processing.html
+	// https://dev.mysql.com/doc/refman/5.7/en/insert-optimization.html
+	// https://dev.mysql.com/doc/refman/5.7/en/optimizing-innodb-bulk-data-loading.html
+	//
+	// Given that there are no UNIQUE constraints in label_membership (well,
+	// apart from the primary key columns), no AUTO_INC column and no FOREIGN
+	// KEY, there is no obvious setting to tweak (based on the recommendations of
+	// the third link above).
+	//
+	// However, in label_membership, updated_at defaults to the current timestamp
+	// both on INSERT and when UPDATEd, so it does not need to be provided.
+
+	runInsertBatch := func(batch [][2]uint) error {
+		sql := `INSERT INTO label_membership (label_id, host_id) VALUES `
+		sql += strings.Repeat(`(?, ?),`, len(batch))
+		sql = strings.TrimSuffix(sql, ",")
+		sql += ` ON DUPLICATE KEY UPDATE updated_at = VALUES(updated_at)`
+
+		vals := make([]interface{}, 0, len(batch)*2)
+		for _, tup := range batch {
+			vals = append(vals, tup[0], tup[1])
+		}
+		return ds.AdhocRetryTx(ctx, func(tx sqlx.ExtContext) error {
+			_, err := tx.ExecContext(ctx, sql, vals...)
+			return errors.Wrap(err, "insert into label_membership")
+		})
+	}
+
+	runDeleteBatch := func(batch [][2]uint) error {
+		rest := strings.Repeat(`UNION ALL SELECT ?, ? `, len(batch)-1)
+		sql := fmt.Sprintf(`
+    DELETE
+      lm
+    FROM
+      label_membership lm
+    JOIN
+      (SELECT ? label_id, ? host_id %s) del_list
+    ON
+      lm.label_id = del_list.label_id AND
+      lm.host_id = del_list.host_id`, rest)
+
+		vals := make([]interface{}, 0, len(batch)*2)
+		for _, tup := range batch {
+			vals = append(vals, tup[0], tup[1])
+		}
+		return ds.AdhocRetryTx(ctx, func(tx sqlx.ExtContext) error {
+			_, err := tx.ExecContext(ctx, sql, vals...)
+			return errors.Wrap(err, "delete from label_membership")
+		})
+	}
+
+	insertBatch := make([][2]uint, 0, insertBatchSize)
+	deleteBatch := make([][2]uint, 0, deleteBatchSize)
+	for _, key := range keys {
+		ins, del, err := getKeyTuples(key)
+		if err != nil {
+			return err
+		}
+		insertBatch = append(insertBatch, ins...)
+		deleteBatch = append(deleteBatch, del...)
+
+		if len(insertBatch) >= insertBatchSize {
+			if err := runInsertBatch(insertBatch); err != nil {
+				return err
+			}
+			insertBatch = insertBatch[:0]
+		}
+		if len(deleteBatch) >= deleteBatchSize {
+			if err := runDeleteBatch(deleteBatch); err != nil {
+				return err
+			}
+			deleteBatch = deleteBatch[:0]
+		}
+	}
+
+	// process any remaining batch that did not reach the batchSize limit in the
+	// loop.
+	if len(insertBatch) > 0 {
+		if err := runInsertBatch(insertBatch); err != nil {
+			return err
+		}
+	}
+	if len(deleteBatch) > 0 {
+		if err := runDeleteBatch(deleteBatch); err != nil {
+			return err
+		}
+	}
+
+	return nil
```

Comment:
Just realized I forgot to update the `label_updated_at` of the host once it has been updated from redis. Adding it now...